### PR TITLE
Logging to FCB2

### PR DESCRIPTION
**Pull request by mkiiskila**
log_fcb_storage_info() nor watermarks yet working.

### CODE REVIEW

File: sys/log/full/src/log_fcb2.c [hunk 1]
```diff
@@ -0,0 +1,866 @@
...
+        }
+    }
+
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+#if MYNEWT_VAL(LOG_PERSIST_WATERMARK)
+    /* Set watermark to first element */
+    memset(&loc, 0, sizeof(loc));
+
+    if (fcb_getnext(fcb, &loc)) {
+        fl->fl_watermark_off = loc.fe_area->fa_off + loc.fe_elem_off;
```
Comment (kasjer) [inline sys/log/full/src/log_fcb2.c]: filed **fe_area** and **fe_elem_off** not present in **FCB2** build

File: sys/log/full/src/log_fcb2.c [hunk 2]
```diff
@@ -0,0 +1,866 @@
...
+     * even possible?) we will never use free space before it thus that space
+     * can be also considered used.
+     */
+    el_min = fcb->f_oldest->fa_off;
+
+    /* Calculate end location of last entry */
+    el_max = fcb->f_active.fe_area->fa_off + fcb->f_active.fe_elem_off;
+
+    /* Sectors assigned to FCB are guaranteed to be contiguous */
+    fa = &fcb->f_sectors[0];
```
Comment (kasjer) [inline sys/log/full/src/log_fcb2.c]: filed **f_sectors** not present in **FCB2** build

File: sys/log/full/src/log_fcb2.c [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+{
+    struct fcb_entry *loc;
+    struct fcb_log *fl;
+    struct log_entry_hdr ueh;
+    int rc;
+
+    loc = (struct fcb_entry*)dptr;
+    fl = (struct fcb_log *)log->l_arg;
+
+    rc = log_fcb_read(log, loc, &ueh, 0, sizeof(ueh));
```
Comment (kasjer) [inline sys/log/full/src/log_fcb2.c]: **log_fcb_read** not present in **FCB2** build

File: fs/fcb2/src/fcb_getprev.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                return FCB_ERR_NOVAR;
+            } else {
+                loc->fe_entry_num = last_valid;
+                rc = fcb_elem_info(loc);
+                assert(rc == 0); /* must be; just succeeded a bit earlier */
+                return rc;
+            }
+        }
+        loc->fe_entry_num++;
+    } while (1);
```
Comment (kasjer) [inline fs/fcb2/src/fcb_getprev.c]: do { } while (1) is less frequent in mynewt then just while (1) { }. This PR already uses while (1) {} in other commit. Nothing wrong just a glitch in the Matrix.

Comment (mkiiskila) [reply fs/fcb2/src/fcb_getprev.c]: Indeed. Obviously code must've started as something else. Will change this.

File: sys/log/full/src/log_fcb2.c [hunk 5]
```diff
@@ -0,0 +1,890 @@
...
+
+    return (0);
+done:
+    return (rc);
+}
+#endif
+
+/**
+ * Copies one log entry from source fcb to destination fcb
+ * @param src_fcb, dst_fcb
```
Comment (kasjer) [inline sys/log/full/src/log_fcb2.c]: quite unusual doxygen comment, usually @param refers to just one argument, other arguments are missing, same for the next function.

File: sys/log/full/src/log_fcb2.c [hunk 6]
```diff
@@ -0,0 +1,890 @@
...
+
+    /* If there are no entries to read and the watermark has not been set */
+    if (fl->fl_watermark_off == 0xffffffff) {
+        fl->fl_watermark_sec = fcb->f_oldest_sec;
+        fl->fl_watermark_off = 0;
+    }
+
+    return (0);
+done:
+    return (rc);
```
Comment (kasjer) [inline sys/log/full/src/log_fcb2.c]: in most cases this file uses just return rc;

File: fs/fcb2/src/fcb_getprev.c (hunk 4 above)
Comment (kasjer) [reply fs/fcb2/src/fcb_getprev.c]: there were two places :), the other one get the fix

File: sys/log/full/include/log/log_fcb.h [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+ *
+ * FCB rotation invalidates all bookmarks.  It is up to the client code to
+ * clear a log's bookmarks whenever rotation occurs.
+ */
+
+/**
+ * @brief Configures an fcb_log to use the specified buffer for bookmarks.
+ *
+ * @param fcb_log               The log to configure.
+ * @param bmarks                The buffer to use for bookmarks.
```
Comment (kasjer) [inline sys/log/full/include/log/log_fcb.h]: bmarks should be buf ?

File: sys/log/full/src/log_fcb2.c [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+        }
+    } while (rc == 0);
+
+    return (rc);
+}
+
+/**
+ * Flushes the log while restoring specified number of entries
+ * using image scratch
+ * @param src_fcb, dst_fcb
```
Comment (kasjer) [inline sys/log/full/src/log_fcb2.c]: params do not match comment, which is also enigmatic 'specified number of entries'.
It was copied from log_fcb.c and that place could also be updated, but since it was not part of this change I did not put comment there.

File: sys/log/full/src/log_fcb2.c [hunk 9]
```diff
@@ -0,0 +1,866 @@
...
+        if (fcb_log->fl_entries) {
+            rc = log_fcb2_rtr_erase(log, fcb_log);
+            if (rc) {
+                goto err;
+            }
+            continue;
+        }
+
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+        old_fa = fcb->f_oldest;
```
[the remark this answers is not in the document]
Comment (mkiiskila) [reply sys/log/full/src/log_fcb2.c]: Watermarking was not yet supported.

File: sys/log/full/src/log_fcb2.c (hunk 1 above)
Comment (mkiiskila) [reply sys/log/full/src/log_fcb2.c]: Watermarking was not yet supported.

File: sys/log/full/src/log_fcb2.c (hunk 2 above)
Comment (mkiiskila) [reply sys/log/full/src/log_fcb2.c]: Watermarking was not yet supported.

File: sys/log/full/src/log_fcb2.c [hunk 10]
```diff
@@ -0,0 +1,890 @@
...
+                    info->used_unread += fcb->f_active.fe_data_off;
+                } else {
+                    info->used_unread +=
+                      fcb->f_active.fe_data_off - fl->fl_watermark_off;
+                }
+#endif
+            }
+        }
+    }
+    if (fl->fl_watermark_sec == FCB_SECTOR_OLDEST) {
```
[the remark this answers is not in the document]
Comment (mkiiskila) [reply sys/log/full/src/log_fcb2.c]: Thanks.

File: sys/log/full/src/log_fcb2.c [hunk 11]
```diff
@@ -0,0 +1,890 @@
...
+    uint8_t buf[LOG_BASE_ENTRY_HDR_SIZE + LOG_IMG_HASHLEN +
+                LOG_FCB2_MAX_ALIGN - 1];
+    struct fcb_entry loc;
+    const uint8_t *u8p;
+    int hdr_alignment;
+    int chunk_sz;
+    int rc;
+    uint16_t hdr_len;
+
+    hdr_len = 0;
```
[the remark this answers is not in the document]
Comment (mkiiskila) [reply sys/log/full/src/log_fcb2.c]: That's true. There's few other occasions like this.

File: sys/log/full/include/log/log_fcb.h (hunk 7 above)
Comment (mkiiskila) [reply sys/log/full/include/log/log_fcb.h]: Yeah, seems likely.

File: sys/log/full/src/log_fcb2.c (hunk 8 above)
Comment (mkiiskila) [reply sys/log/full/src/log_fcb2.c]: Yup, the comments don't appear to match the code.
